src/wayback/_http.py

```diff
@@ -0,0 +1,282 @@
+
+    def close(self, cache: bool = True) -> None:
+        with self._read_lock:
+            # Reading bytes potentially involves decoding data from compressed
+            # gzip/brotli/etc. responses, so we need to handle those errors by
+            # continuing to just read the raw data off the socket instead.
+            #
+            # This fallback behavior is inspired by requests:
+            #   https://github.com/psf/requests/blob/eedd67462819f8dbf8c1c32e77f9070606605231/requests/sessions.py#L160-L163
+            # For urllib3, the appropriate errors to handle would be:
+            #   `(DecodeError, ProtocolError, RuntimeError)`
+            try:
+                if cache:
+                    try:
+                        self.content
+                    except (ChunkedEncodingError, ContentDecodingError, RuntimeError):
+                        self._raw.raw.read(decode_content=False)
+                else:
+                    self._raw.raw.read(decode_content=False)
+            finally:
+                self._raw.close()
+
+
+class WaybackHttpAdapter:
+    """
+    Handles making actual HTTP requests over the network. For now, this is an
+    implementation detail, but it may be a way for users to customize behavior
+    in the future.
+    """
+
+    def __init__(self) -> None:
+        self._session = requests.Session()
+
+    # FIXME: remove `allow_redirects`! Redirection needs to be handled by
+    # whatever does throttling, which is currently not here.
+    def request(
+        self,
+        method: str,
+        url: str,
+        *,
+        params: dict = None,
+        headers: dict = None,
+        allow_redirects: bool = True,
+        timeout: Union[int, Tuple[int, int]] = None
+    ) -> WaybackHttpResponse:
+        """
+        Send an HTTP request and return a :class:`WaybackHttpResponse` object
+        representing the response.
+
+        Parameters
+        ----------
+        method : str
+            Method to use for the request, e.g. ``'GET'``, ``'POST'``.
+        url : str
+            The URL to reqeust.
+        params : dict, optional
+            For POST/PUT requests, data to be encoded in the response body. For
+            other methods, this should be encoded as the querystring.
+        headers : dict, optional
+            The HTTP headers to send with the request.
+        allow_redirects : bool, default: True
+            Whether to follow redirects before returning a response.
+        timeout : int or float or tuple of (int or float, int or float), optional
+            How long to wait, in seconds, before timing out. If this is a single
+            number, it will be used as both a connect timeout (how long to wait
+            for the first bit of response data) and a read timeout (how long
+            to wait between bits of response data). If a tuple, the values will
+            be used as the connect and read timeouts, respectively.
```

Comment:
I’m a little worried this is too narrowly designed around how requests does it. It's probably not a huge deal, though, and keeps the timeout interface the same-ish as it is today.